src/datasource/gateway.js

```diff
@@ -0,0 +1,273 @@
+      url: '/v1/query/apps',
+      method: 'post',
+      data: {
+        height,
+        opts: {
+          staking_status,
+          blockchain,
+          page,
+          per_page,
+        }
+      }
+    }),
+    getSupply: (height) => ({
+      url: '/v1/query/supply',
+      method: 'post',
+      data: {
+        height,
+      }
+    }),
+    getNodes: (staking_status, jailed_status, height, blockchain, page, per_page) => ({
+      url: '/v1/query/nodes',
+      method: 'post',
+      data: {
+        height,
+        opts: {
+          staking_status,
+          jailed_status,
+          blockchain,
+          page,
+          per_page,
+        }
+      }
+    }),
+  }
+
+  use(provider) {
+    this.provider = provider;
+    return this;
+  }
+
+  // looks ugly with the ifs
+  // but the gateway constantly responds with 200
+  // and responds with errors in response.data
+  // in a non-consistent form.
+  parseSuccessfulResponse = (response) => {
+    
+    if (typeof response.data === 'string' && response.data.indexOf('Method Not Allowed') > -1) {
+      throw {
+        message: 'Method Not Allowed',
+      }
+    }
+
+    if (response.data.code && response.data.code !== 200) {
+      throw response.data;
+    }
+
+    return response.data;
+  }
+
+  parseErrorResponse = (error) => {
+    if (error.response && error.response.data && error.response.data.error) {
+      throw error.response.data.error;
+    }
+
+    if (typeof error === 'string') {
+      throw {
+        message: error,
+      }
+    }
+
+    throw error;
+
+  }
+
+  perform = async (requestName, ...args) => {
+    const reqConfig = typeof this.requests[requestName] === 'function' ?
+      this.requests[requestName](...args) :
+      this.requests[requestName];
+
+    const response = await this
+      .provider
+      .http
+      .request(reqConfig)
+      .then(this.parseSuccessfulResponse)
+      .catch(this.parseErrorResponse);
+
+    return response;
+  }
+
+  // does not really need to be bound to `this`, but keeping it for semantics' sake.
+  // arguments explicit forward for clear signature lookup, avoid using `...args` 
+  _getHeight = this.perform.bind(this, 'getHeight');
+  _getAccount = (id) => this.perform.call(this, 'getAccount', id);
+  _getTransaction = (id) => this.perform.call(this, 'getTransaction', id);
+  _getBlock = (height) => this.perform.call(this, 'getBlock', height);
+  _getBlockTxs = (height, prove, page, per_page) => this.perform.call(this, 'getBlockTxs', height, prove, page, per_page);
+  _getApps = (staking_status, height, blockchain, page, per_page) => this.perform.call(this, 'getApps', staking_status, height, blockchain, page, per_page);
+  _getSupply = this.perform.bind(this, 'getSupply');
+  _getNodes = (staking_status, jailed_status, height, blockchain, page, per_page) => this.perform.call(this, 'getNodes', staking_status, jailed_status, height, blockchain, page, per_page);
```

Comment:
Do we really need this though? It's kind of ugly—JS classes are good enough by default.

Reply:
Is it needed as of now? Not very much, but that's a placeholder for proxies or middlewares to be easily inserted per request if needed, think specific endpoint request/response transformers.

Underscores look ugly and feel slashy as if we are in C++, I get it.

Reply:
Haha yeah, it's mostly a matter of style... but what you say makes sense. :)

Reply:
I think the wallet will have a use-case for this where it is more obvious

Reply:
Yup, for sure! Resolving this, I think it looks good